feat: тесты p2p-request-ов

alinkalina opened the pull request:
Пока что:

- **tests\test_p2p_request.py** - добавлена проверка авторизации и scope-a 'p2p-requst' у пользователя, создающего p2p-request и запрашивающего ревью

- из **tests\test_users.py** перенесла создание scopes в новый файл, потому что он выполняется первым (я так понимаю, файлы с тестами запускаются по алфавиту, и test_p2p_request.py самый первый, так что решила создавать scopes именно там)

alinkalina committed:
feat: тесты создания p2p-request-a

Добавлена проверка авторизации и scope-a 'p2p-requst' у пользователя, создающего p2p-request и запрашивающего ревью - tests\test_p2p_request.py.

Из tests\test_users.py перенесла создание scopes в новый файл, потому что он выполняется первым.

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -0,0 +1,104 @@
+        with UserAccessCookie(client, test_p2p_user.username, ''):
+            response = self.do_request(params={
+                'username': 'test_incorrect_scope_user',
+                'password': 'test_incorrect_scope_user',
```
А вот здесь лучше использовать действительные данные для авторизации, чтобы мы проверяли именно работу API со scopes (а то сейчас у нас может быть неправильный username/password и будет та же ошибка, что при неправильных scopes)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -0,0 +1,104 @@
+
+        assert response.status_code == 401
+
+        assert users_before_request == len(UserCrud(db).get_many())
```
`UserCrud(db)` используется дважды, поэтому лучше вынести в отдельную переменную вроде: `user_crud = UserCrud(db)`. (и да, это косяк мой из ранее созданных тестов)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -0,0 +1,104 @@
+        with UserAccessCookie(client, test_p2p_user.username, ''):
+            response = self.do_request(params={
+                'username': 'test_incorrect_scope_user',
+                'password': 'test_incorrect_scope_user',
```
> А вот здесь лучше использовать действительные данные для авторизации, чтобы мы проверяли именно работу API со scopes (а то сейчас у нас может быть неправильный username/password и будет та же ошибка, что при неправильных scopes)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -0,0 +1,104 @@
+
+        assert response.status_code == 401
+
+    def test_incorrect_scope(self):
```
Содержание метода не отвечает его семантике. То есть по названию класса и метода складывается ощущение, что он должен тестировать взаимодействие с p2p request route, при неправильных scopes, а по факту тут вообще не это происходит (авторизовать пользователя можно через `UserAccessCookie`)

LaGGgggg reviewed (changes requested):
В целом логика того, как это должно выглядеть уловлена, скоро будет всё красиво и круто.

LaGGgggg commented:
> * из **tests\test_users.py** перенесла создание scopes в новый файл, потому что он выполняется первым (я так понимаю, файлы с тестами запускаются по алфавиту, и test_p2p_request.py самый первый, так что решила создавать scopes именно там)

Мысль неплохая, крутой research, но лучше уж тогда вынести создание scopes на уровень выше. Тебе, наверное, уже не стоит с этим бодаться, позже сам перенесу куда-нибудь. В целом лайк

alinkalina committed:
feat: тесты создания ревью p2p-request-a

test_user_gets_his_own_project - проверяет, что пользователю не может быть выдан его же проект
test_correct - проверяет, что при начале ревью review_state меняется
test_user_already_have_review - проверяет, что у юзера может быть только одно ревью
test_no_pending_projects - проверяет, что если не начатых ревью нет, пользователю выдаётся соответствующее сообщение

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -90,15 +108,59 @@ def test_not_authorized(self):
+
+        p2p_request_schema = schemas.P2PRequest.model_validate(test_p2p_request)
+
+        assert p2p_request_schema.model_dump()['review_state'] != response.json()['review_state']
```
Не понял зачем эта проверка, догадки есть, но гадать не хочу. Можно пояснительную бригаду?

alinkalina replied:
Это проверяет что у проекта изменяется pending на progress. Не уверена, что это нужно, но т к у меня это получилось, решила отставить)) Почему бы и нет?

LaGGgggg replied:
Имеет смысл, но тогда уж нужно проверять именно что изменилось на pending, а не что отличное от review_state в схеме. Плюс лучше в базу данных залезть и проверить, что там всё действительно поменялось. НО, для test_p2p_request нужно сначала заново загрузить данные из бд через refresh (так как мы данные в бд изменили, но эти данные из бд не подтянули)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -90,15 +108,59 @@ def test_not_authorized(self):
+
+        assert p2p_request_schema.model_dump()['review_state'] != response.json()['review_state']
+
+        assert schemas.P2PRequest.model_validate(oldest_project).id == response.json()['id'] + 1
```
И смысл этого тоже не очень понял, если честно. Тут, видимо, какая-то хитроумная проверка на правильность выдачи нужного request-а, но я не понимаю

alinkalina replied:
Вот именно что хитроумная... В общем, здесь я пыталась сначала найти подходящий (самый старый) p2p-request и сравнить его id с тем, который в итоге получил юзер (думала, они будут равны). НО оказалось, что их id отличаются на 1 (это можно даже в таблице увидеть, если запустить). Вот я и сделала такое сравнение, хотя тоже не до конца поняла, почему происходит именно так.

LaGGgggg replied:
start_review же новый review создаёт, а не берёт его из базы данных, поэтому id и отличается. Нужно в базу данных обратиться и через crud получить самый старый объект (с самым большим id, по идее)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -90,15 +108,59 @@ def test_not_authorized(self):
+
+        assert schemas.P2PRequest.model_validate(oldest_project).id == response.json()['id'] + 1
+
+    def test_user_already_have_review(self):
```
Я так понял, что этот тест фактически зависит от предыдущего. Если это действительно так, то это плохая практика, потому что может стать причиной трудноотлавливаемых ошибок из-за порядка запуска тестов. Тесты лучше делать независимыми друг от друга (если это возможно). Предыдущий тест может не запуститься при определённых обстоятельствах, он может сломаться и не сделать того, что нужно для этого теста и т.д.
P.S. У меня был личный неприятный опыт отлова ошибок, которые возникали из-за порядка запуска тестов, такое лучше не ловить

alinkalina replied:
Даа, а я так надеялась что так можно... Ну ладно, значить для каждого теста лучше создать независимые переменные, которые не будут изменяться в соседних тестах?

LaGGgggg replied:
Можно использовать и глобальные, но необходимо добиться состояния, когда вне зависимости от того, что произошло в этом тесте, ничто не поменяется вне него. То есть всё равно запустился этот тест или нет, прошёл он или нет, вот это идеал (инкапсуляция, к слову)

LaGGgggg commented on API/tests/test_p2p_request.py:
```diff
@@ -90,15 +108,59 @@ def test_not_authorized(self):
+
+        assert response.json() == {'context': 'You already have a review, complete it first'}
+
+    def test_no_pending_projects(self):
```
Аналогичный alert, что и с тестом выше

LaGGgggg reviewed (commented):
Придраться особо не к чему, кроме одного теста, который я не понял и предупреждений относительно независимости тестов, all good at all
